README.md
# MMM-UKMOWeatherWarnings

This is a module for the [MagicMirror²](https://github.com/MichMich/MagicMirror/).

This module displays Weather Warnings from the UK Met Office.

![Weather warnings](warnings.png "Weather warnings.") OR ![Weather warnings](warnings-bw.png "Weather warnings black and white.")

This module will also send notifications of Weather Warnings for use by other modules.

## Using the module

### Installation
1. Navigate into your MagicMirror's `modules` folder and execute `git clone https://github.com/maloakes/MMM-UKMOWeatherWarnings`.  A new folder `MMM-UKMOWeatherWarnings` will appear.

### Config
To use this module, add a configuration block, similar to this example, to the modules array in the `config/config.js` file:
```js
var config = {
    modules: [
        {
            module: 'MMM-UKMOWeatherWarnings',
            config: {
                // See below for configurable options
                region: 'sw',
                apiBase: 'http://www.metoffice.gov.uk/public/data/PWSCache/WarningsRSS/Region/',
        }
    ]
}
```

## Configuration options

| Option              | Description
|-------------------- |-----------
| `apiBase`           | *Required* UKMO Weather Warnings feed URL <br><br>**Type:** `string` <br>e.g. `http://www.metoffice.gov.uk/public/data/PWSCache/WarningsRSS/Region/`
| `region`            | *Required* UKMO Weather Warnings Region <br><br>**Type:** `string` <br>See table below, e.g. `sw`
| `updateInterval`    | *Optional* Update interval <br><br>**Type:** `int`(milliseconds) <br>Default 900000 milliseconds (15 minutes)
| `retryDelay`        | *Optional* Retry delay <br><br>**Type:** `int`(milliseconds) <br>Default 5000 milliseconds (5 seconds)
| `tableClass`        | *Optional* Size used for table <br><br>**Type:** `string` <br>Default `small`
| `colored`           | *Optional* Whether to color the warning symbol or display level as character <br><br>**Type:** `boolean` <br>Default `true`
| `broadcastCurrent`  | *Optional* Whether to send notifications for current warnings <br><br>**Type:** `boolean` <br>Default `false`
| `broadcastForecast` | *Optional* Whether to send notifications for warnings during forecast period <br><br>**Type:** `boolean` <br>Default `false`
| `broadcastListSize` | *Optional* Size (i.e. number of days) of the forecast list <br><br>**Type:** `int` <br>Default 5
| `useTestData`       | *Optional* Whether to use test data instaed of the live feed <br><br>**Type:** `boolean` <br>Default `false`
| `testUrl`           | *Optional* URL of the test data file  <br><br>**Type:** `string` <br>e.g. `http://localhost:8080/modules/MMM-UKMOWeatherWarnings/public/warns.json`
| `backgroundMode`    | *Optional* Whether to run in background mode, only sending notifications <br><br>**Type:** `boolean` <br>Default `false`
| `delta`             | *Optional* Look ahead delta for current warnings <br><br>**Type:** `int`(minutes) <br>0

#### Regions
| Code | Region
|------|---------------------
| os   | Orkney & Shetland
| ta   | Central, Tayside & Fife
| he   | Highlands & Eilean Siar
| gr   | Grampian
| st   | Strathclyde
| dg   | Dumfries, Galloway, Lothian & Borders
| ni   | Northern Ireland
| wl   | Wales
| nw   | North West England
| ne   | North East England
| yh   | Yorkshire & Humber
| wm   | West Midlands
| em   | East Midlands
| ee   | East of England
| sw   | South West England
| se   | London & South East England

## Notifications
This module can send notifications of weather warnings.
The module can be run in background mode to send the notifications without displaying the data itself.

Example payload item;
`{ level: "A", color: "orange", type: "Snow" }`

| Notification               | Payload
|----------------------------|-----------------------------------------
| `WEATHER-WARNING-CURRENT`  | Highest level alert for the current period
| `WEATHER-WARNING-FORECAST` | Array of highest level alert for each day of forecast period

## Test Data
Test data can be generated, since we don't get warnings all of the time.

1. Navigate into your MagicMirror's `modules/MMM-UKMOWeatherWarnings/testing` folder.
2. Execute `node createDummyWarnings.js`
3. A file, `warns.json` will be generated in your modules `public` folder
4. Edit your config.js file to set `useTestData` and `testUrl`
5. Start/Restart your mirror


CHANGELOG.md
# MMM-UKMOWeatherWarnings Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).


## [1.0.0] - Unreleased

First public release


testing/createDummyWarnings.js
/* Magic Mirror
 * Module: MMM-UKMOWeatherWarnings
 *
 * By Malcolm Oakes https://github.com/maloakes
 *
 * MIT Licensed.
 *
 * create dummy warnings data in the current timeframe
 *
 * the JSON file can be loaded by setting the module config data
 */

const moment = require('moment')
const fs = require('fs')

let warns = createDummyData()
let fpath = '../public/warns.json'
writeFile(fpath, warns)

/*
 * create the data for the current time frame
 */
function createDummyData() {
  let fmt = 'ddd DD MMM'

  let hObj = createWarnHeader()

  let wTime = moment()
  let wfrom = '0200 ' + wTime.format(fmt)
  let wto = '1800 ' + wTime.format(fmt)
  hObj.items.push(createWarnItem('Red', 'Fog', wfrom, wto))

  wTime = moment()
  wfrom = '1600 ' + wTime.format(fmt)
  wto = '1800 ' + wTime.add(1, 'days').format(fmt)
  hObj.items.push(createWarnItem('Yellow', 'Snow and Ice', wfrom, wto))

  wTime = moment().add(1, 'days')
  wfrom = '0800 ' + wTime.format(fmt)
  wto = '1900 ' + wTime.add(1, 'days').format(fmt)
  hObj.items.push(createWarnItem('Amber', 'Rain', wfrom, wto))

  wTime = moment().add(1, 'days')
  wfrom = '0000 ' + wTime.format(fmt)
  wto = '1000 ' + wTime.format(fmt)
  hObj.items.push(createWarnItem('Red', 'Fog', wfrom, wto))

  return hObj

}

/*
 * create header, based on real data with times in current frame
 */
function createWarnHeader() {

  let now = moment.utc()
  let fmtTtl = 'DD-MMM-YYYY HH:mm'
  let fmtPub = 'ddd, DD MMM YYYY HH:mm:ss'
  let fmtDate = 'YYYY-MM-DDTHH:mm:ss'

  let hObj = {
      type: 'rss 2.0',
      title: 'Met Office Warnings - dummy data generated ' + now.format(fmtTtl),
      link: 'https://www.metoffice.gov.uk/public/weather/warnings/?regionName=gr',
      description: 'Weather Warnings of severe and extreme weather from the Met Office',
      language: 'en-gb',
      copyright: '(c) Crown copyright',
      pubdate: now.format(fmtPub) + ' GMT',
      'dc:creator': 'webteam@metoffice.gov.uk',
      'dc:date': now.format(fmtDate) + 'Z',
      'dc:language': 'en-gb',
      'dc:rights': '(c) Crown copyright',
      image:
       { text: '\r\n      \r\n      \r\n      \r\n    ',
         title: 'Met Office Warnings for Grampian',
         url: 'https://www.metoffice.gov.uk/lib/template/logo_crop.gif',
         link: 'https://www.metoffice.gov.uk/public/weather/warnings/?regionName=gr' },
      items: []
  }
  return hObj
}

/*
 * create item, based on real data with times in current frame
 */
 function createWarnItem(level, type, from, to) {
  var witemObj = {
    title: level + ' Warning of ' + type + ' affecting Scotland (Grampian)',
    link: 'https://www.metoffice.gov.uk/public/weather/warnings/?regionName=gr&from=rss&sn=bb9e7f32-549d-490e-bf5c-6f5fdef26723',
    description: 'Red Warning of Fog affecting Scotland (Grampian) : Aberdeen, Aberdeenshire, Moray ' +
                  'valid from ' + from + ' to ' + to,
    'metadata:warningkind': 'WARNING',
    'metadata:warninglevel': level.toUpperCase(),
    'metadata:warningtype': type.toUpperCase()
  }

  return witemObj
}

/*
 * write the JSON file
 */
function writeFile(fpath, wobj) {
  fs.writeFileSync(fpath, JSON.stringify(wobj))
}


node_helper.js
/* Magic Mirror
 * Node Helper: MMM-UKMOWeatherWarnings
 *
 * By Malcolm Oakes https://github.com/maloakes
 *
 * MIT Licensed.
 */
const FeedMe = require('feedme')
const http = require('http')
const moment = require('moment')
const fs = require('fs')
var NodeHelper = require("node_helper");

module.exports = NodeHelper.create({

	// Subclass start method.
	start: function() {
		console.log("Started node_helper.js for MMM-UKMOWeatherWarnings.");
	},

	// Override socketNotificationReceived method.

	/* socketNotificationReceived(notification, payload)
	 * This method is called when a socket notification arrives.
	 *
	 * argument notification string - The identifier of the noitication.
	 * argument payload mixed - The payload of the notification.
	 */
	socketNotificationReceived: function(notification, payload) {

		if (notification === "MMM-UKMOWeatherWarnings-WARNINGS-REQ") {
			this.warningsRequest(payload)
		}
	},

	/*
	 * request for the warning message
	 * get the message from the feed
	 * notify the module
	 */
	warningsRequest: function(url) {
		self = this

		http.get(url, (res) => {
		  if (res.statusCode != 200) {
		    console.error(new Error(`status code ${res.statusCode}`))
		    return
		  }
		  var parser = new FeedMe(true)

		  res.pipe(parser)
		  parser.on('end', () => {
		    res = parser.done()
				self.sendSocketNotification('MMM-UKMOWeatherWarnings-WARNINGS-RESULT', res);

		  })
		})
	},

});


warningsobject.js
/* Magic Mirror
 * WarningsObject: MMM-UKMOWeatherWarnings
 *
 * By Malcolm Oakes https://github.com/maloakes
 *
 * MIT Licensed.
 *
 * This class represents Weather Warning message data from the UK Met Office.
 * This will contain 0 or more WarningItemObjects.
 */

 class WarningsObject {
	constructor(warningsResult) {

		this.title = warningsResult.title
		this.items = null
		this.ordering = null

		// build the Warning Item ordering map
		this.buildOrdering()

		// create the list of Warning items
		var itemList = []
		for (var i in warningsResult.items) {
			const item = new WarningItemObject(warningsResult.items[i])
			// only add the item is valid in the future
			if (item.isFuture()) {
			  itemList.push(item)
			}
		}

		// sort the list if it contains objects
		if (itemList.length > 0) {
			this.items = itemList.sort(function(a,b){return a.validFrom.diff(b.validFrom)})
		}
	}

  /*
	 * get colour (severity) of item
	 */
	getColor(idx) {
		return this.items[idx].getColor()
	}

	/*
	 * get type of item
	 */
	getTypeTc(idx) {
		return this.items[idx].getTypeTc()
	}

	/*
	 * get validity of item
	 */
	getValidity(idx) {
		this.items[idx].getValidity()
	}

	/*
	 * get the maximum current warning
	 */
	getMaxCurrentWarn(delta) {
		// get warnings that are now current
		// return level and info of highest level

		let ordArray = this.getMaxPeriodWarn('current', delta)

		// just return level and type (title case)
		let res = {}
		if (ordArray) {
		    res[0] = ordArray.length > 0 ? { level: ordArray[0].getLevelCode(),
			     color: ordArray[0].getColor(),
			     type: ordArray[0].getTypeTc() } : null
		}
		return res
	}

  /*
	 * determine the warnings for the required period
	 * parameter:
	 *  mode: current - active now
	 *        today   - active from now to midnight
	 *        day     - active on day specified by offset
	 *  offset:
	 *    mode = 'current' - offset (in minutes) from now
	 *    mode = 'day' - offset (in days) from today
	 */
	getMaxPeriodWarn(mode, offset) {

	    if (this.items) {

		let newArray = null

		switch (mode) {
			case 'current':
				// get warnings that are now current
				// return level and info of highest level
			  newArray = this.items.filter(function (el) {
		  		return el.validFrom.isBefore(moment().add(offset, 'minutes')) && el.validTo.isAfter(moment())
				});
				break;
			case 'today':
				// get warnings active from now until end of day
				// return highest
				// Vf < Ed & Vt > Nw
				newArray = this.items.filter(function (el) {
				  return el.validFrom.isBefore(moment().endOf('day')) && el.validTo.isAfter(moment())
				});
				break;
			case 'day':
				// get warnings active during the day
				// return highest
				// Vf < Ed & Vt > Sd
				newArray = this.items.filter(function (el) {
				  return el.validFrom.isBefore(moment().add(offset, 'days').endOf('day')) && el.validTo.isAfter(moment().add(offset, 'days').startOf('day'))
				});
				break;
		}
		return this.sortByLevel(newArray)
	    } else {
		return null
	    }
	}

	/*
	 * Get the highest level warning for each day of the period
	 */
	getMaxWarnFcList(listSize) {
		// get warnings from now to end of today
		// & get warnings the following days
		// return highest level for each day
		let fcArray = []
		let tmp = this.getMaxPeriodWarn('today', null)
		fcArray.push(tmp && tmp.length > 0 ? tmp[0] : null)

		for (let offset = 1; offset < listSize; offset++) {
		  let tmp = this.getMaxPeriodWarn('day', offset)
		  fcArray.push(tmp && tmp.length > 0 ? tmp[0] : null)
		}

		let res = {}
		for (let i in fcArray) {
		  res[i] = fcArray[i] != null ? { level: fcArray[i].getLevelCode(),
			     color: fcArray[i].getColor(),
			     type: fcArray[i].getTypeTc() } : null
		}

		return res
	}

  /*
	 * sort the list by level
	 */
	sortByLevel(warnArray) {
		let ord = this.ordering
		var ordArray = warnArray.sort( function(a, b) {
				return (ord[a.level] - ord[b.level]) || a.type.localeCompare(b.type);
		});

		return ordArray
	}

	/*
	 * build the ordering map in the constuctor by calling this
	 * The warning items should be sorted in order of severity
	 */
	buildOrdering(warnArray) {
	  this.ordering = {} // map for efficient lookup of sortIndex
	  let sortOrder = ['RED','AMBER','YELLOW'];
	  for (let i=0; i<sortOrder.length; i++) {
	    this.ordering[sortOrder[i]] = i;
	  }
	}

}


warningitemobject.js
/* Magic Mirror
 * WarningObjectItem: MMM-UKMOWeatherWarnings
 *
 * By Malcolm Oakes https://github.com/maloakes
 *
 * MIT Licensed.
 *
 * This class represents an item from a Weather Warning message data from the UK Met Office.
 */

 class WarningItemObject {
	constructor(warningItem) {

		this.validFrom = null
		this.validTo = null
		this.title = null
    this.description = null
		this.type = null
		this.kind = null
    this.level = null
		this.region = null

		this.parseItem(warningItem)
	}


	/*
	 * is this warning item in the isFuture?
	 * i.e. after now
	 */
  isFuture() {
    return this.validTo.isAfter()
  }

	/*
	 * parse the JSON to populate this object
	 */
  parseItem(item) {
    this.title = item.title
    this.description = item.description
    let validity = this.parseTimes(this.description)
    this.validFrom = validity[0]
    this.validTo = validity[1]
    this.region = this.parseRegion(item.link)
    this.level = item['metadata:warninglevel']
    this.kind = item['metadata:warningkind']
    this.type = item['metadata:warningtype']
  }

	/*
	 * extract the region code
	 */
  parseRegion(link) {
    let rg = link.indexOf('regionName=')
    const region = link.substr(rg + 11, 2)

    return region
  }

	/*
	 * extract the valid from and valid to times
	 */
  parseTimes(description) {
    const times = []

    let vf = description.indexOf('valid from')
    let to = description.lastIndexOf(' to ')
    let fs = description.substr(vf + 11, to - vf - 11)
    let ts = description.substr(to + 4, description.length - to - 4)

    let format = 'HHmm ddd DD MMM'

    let vfTime = moment(fs, format)
    let vtTime = moment(ts, format)

    times.push(vfTime)
    times.push(vtTime)

    // handle crossing the year boundary
    if(vtTime.diff(vfTime) < 0) {
      vtTime.add(1, 'years')
    }

    return times
  }

	/*
	 * get the colour (level) of the warning
	 * translate AMBER to ORANGE for display
	 */
	getColor() {
		return this.level === 'AMBER' ? 'orange' : this.level.toLowerCase()
	}

	/*
	 * extract the type and convert to title case
	 */
	getTypeTc() {
		var str = this.type.toLowerCase().split(' ');
		for (var i = 0; i < str.length; i++) {
			if (str[i] != 'and') {
				str[i] = str[i].charAt(0).toUpperCase() + str[i].slice(1);
			}
		}
		return str.join(' ');
	};

	/*
	 * get the validity time range
	 */
	getValidity() {
		let fmt = 'ddd HH:mm'
		return this.validFrom.format(fmt) + ' - ' + this.validTo.format(fmt)
	}

	/*
	 * get the code for the level by extracting the first character,
	 * i.e. 'R', 'A', 'Y'
	 */
	getLevelCode() {
		return this.level.substr(0, 1)
	}

}


MMM-UKMOWeatherWarnings.js
/* Magic Mirror
 * Module: MMM-UKMOWeatherWarnings
 *
 * By Malcolm Oakes https://github.com/maloakes
 *
 * MIT Licensed.
 *
 * Gets Weather warnings from the UK Met Office feed.
 * These can be dispayed in tabular format.
 * Optionally Notify waring data for use by other modules.
 */

Module.register("MMM-UKMOWeatherWarnings", {
	defaults: {
		updateInterval: 900000, //15 mins
		retryDelay: 5000,
		tableClass: "small",
		apiBase: null,
		region: null,
		colored: true,
		broadcastCurrent: true,
		broadcastForecast: true,
		broadcastListSize: 5, // size of the forecast warnings list to broadcast
		useTestData: false,
		testUrl: null,
		backgroundMode: false,
		delta: 0, //minutes
	},

	requiresVersion: "2.1.0", // Required version of MagicMirror

	start: function() {
		var self = this;
		var dataRequest = null;
		var dataNotification = null;
		var warningsData = null;

		//Flag for check if module is loaded
		this.loaded = false;

		//Do this once first
		self.sendSocketNotification('MMM-UKMOWeatherWarnings-WARNINGS-REQ', this.getUrl());


		// Schedule update timer.
		setInterval(function() {
			self.sendSocketNotification('MMM-UKMOWeatherWarnings-WARNINGS-REQ', self.getUrl());
			self.updateDom();
		}, this.config.updateInterval);
	},

	/*
	 * build the URL string
	 */
	getUrl: function() {
		return this.config.useTestData ? this.config.testUrl : this.config.apiBase + this.config.region
	},

	/* scheduleUpdate()
	 * Schedule next update.
	 *
	 * argument delay number - Milliseconds before next update.
	 *  If empty, this.config.updateInterval is used.
	 */
	scheduleUpdate: function(delay) {
		var nextLoad = this.config.updateInterval;
		if (typeof delay !== "undefined" && delay >= 0) {
			nextLoad = delay;
		}
		nextLoad = nextLoad ;
		var self = this;
		setTimeout(function() {
			self.sendSocketNotification('"MMM-UKMOWeatherWarnings-WARNINGS-REQ', null);
		}, nextLoad);
	},



	getScripts: function() {
		return [
			'warningsobject.js',
			'warningitemobject.js',
		];
	},

	getStyles: function () {
		return [
			"font-awesome.css",
			"MMM-UKMOWeatherWarnings.css",
		];
	},

	getTemplate: function () {
		return "MMM-UKMOWeatherWarnings.njk"
	},

	getTemplateData: function () {
		return {
			config: this.config,
			warnings: this.warningsData
		}
	},

	// Load translations files
	getTranslations: function() {
		return {
			en: "translations/en.json",
			es: "translations/es.json"
		};
	},


	// socketNotificationReceived from helper
	socketNotificationReceived: function (notification, payload) {

		if(notification === "MMM-UKMOWeatherWarnings-WARNINGS-RESULT") {
			// set dataNotification

			this.warningsData = new WarningsObject(payload);
			this.broadcastWarnings()

			this.updateDom();
		}
	},

	/*
	 * broadcast notifications to other modules if required
	 */
	broadcastWarnings: function() {
		if (this.config.broadcastCurrent) {
			var resCurr = this.warningsData.getMaxCurrentWarn(this.config.delta)
			this.sendNotification("WEATHER-WARNING-CURRENT", resCurr)
		}
		if (this.config.broadcastForecast) {
			var resFcList = this.warningsData.getMaxWarnFcList(this.config.broadcastListSize)
			this.sendNotification("WEATHER-WARNING-FORECAST", resFcList)
		}
	},


});
